db/blip_connected_client.go

```diff
@@ -68,3 +76,68 @@ func (bh *blipHandler) handlePutRev(rq *blip.Message) error {
 	}
 	return bh.processRev(rq, &stats)
 }
+
+// Handles a Connected-Client "query" request.
+func (bh *blipHandler) handleQuery(rq *blip.Message) error {
+	if _, found := rq.Properties[QuerySource]; found {
+		return base.HTTPErrorf(http.StatusForbidden, "Only named queries allowed")
+	}
+	name, found := rq.Properties[QueryName]
+	if !found {
+		return base.HTTPErrorf(http.StatusBadRequest, "Missing 'name'")
+	}
+	query, found := bh.db.Options.ConnectedClientQueries[name]
+	if !found {
+		return base.HTTPErrorf(http.StatusNotFound, "No such query '%s'", name)
+	}
+
+	var params map[string]interface{}
+	bodyBytes, err := rq.Body()
+	if err != nil {
+		return err
+	}
+	if len(bodyBytes) > 0 {
+		if err = json.Unmarshal(bodyBytes, &params); err != nil {
+			return base.HTTPErrorf(http.StatusBadRequest, "Invalid query parameters")
+		}
+	}
+	for _, paramName := range query.Parameters {
+		if _, found := params[paramName]; !found {
+			return base.HTTPErrorf(http.StatusBadRequest, "Parameter '%s' is missing", paramName)
+		}
+	}
+
+	bh.logEndpointEntry(rq.Profile(), fmt.Sprintf("name: %s", name))
+
+	// In the N1QL string, replace "BUCKET" with the actual bucket name, wrapped in back-quotes in
+	// case it's not a valid N1QL identifier (e.g. `travel-sample`.)
+	// TODO: This is a workaround for the inability to use back-quotes for this in an SG config
+	// file, because those are already reserved for quoting long literals like the sync function.
+	statement := strings.ReplaceAll(query.Statement, "BUCKET", "`"+bh.db.Name+"`")
+
+	// Run the query:
+	results, err := bh.db.N1QLQueryWithStats(bh.db.Ctx, QueryTypeConnectedClientPrefix+name, statement, params,
+		base.RequestPlus, false)
+	if err != nil {
+		var qe *gocb.QueryError
+		if errors.As(err, &qe) {
+			base.WarnfCtx(bh.loggingCtx, "Error running query %q: %v", name, err)
+			return base.HTTPErrorf(http.StatusInternalServerError, "Query error: %s", qe.Errors[0].Message)
+		} else {
+			base.WarnfCtx(bh.loggingCtx, "Unknown error running query %q: %T %#v", name, err, err)
+			return base.HTTPErrorf(http.StatusInternalServerError, "Unknown error running query")
+		}
+	}
+
+	// Write the results to the response:
+	var out bytes.Buffer
+	enc := json.NewEncoder(&out)
```

Comment:
Util func that wraps different JSON libraries for CE/EE

```suggestion
	enc := base.JSONEncoder(&out)
```